### PR TITLE
feat(stark-ui): implementation of a session timeout warning

### PR DESCRIPTION
**Pull request by Mallikki**
ISSUES CLOSED: #719 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NationalBankBelgium/stark/blob/master/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```
Issue Number: #119 #719 

## What is the new behavior?

A new SessionTimeoutWarning dialog will be displayed when the user is inactive for too long, along with a timer. 

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```
## Other information

### CODE REVIEW

**Comment by Mallikki** [comment]
Hey guys, I still need to look after the translation problem, but I've made the pull request anyway so you can already have a look at the change ;)

File: packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effect.spec.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+			expect(mockObserver.next).toHaveBeenCalledTimes(1);
+			expect(mockObserver.error).not.toHaveBeenCalled();
+			expect(mockObserver.complete).not.toHaveBeenCalled();
+			expect(mockDialogService.closeAll).toHaveBeenCalledTimes(1);
+		});
+	});
+
+	describe("On ngrxOnRunEffects", () => {
+		it("Shouldn't use the effects", () => {
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effect.spec.ts]: Can you add more info in the description... "should stop the effects immediately when the option xxxx is true"

File: packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effect.spec.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+			const mockResolvedEffectsSubject: Subject<any> = new Subject<any>();
+			const mockResolvedEffects$: Observable<any> = mockResolvedEffectsSubject.asObservable();
+
+			const resolvedEffectsObservable: Observable<EffectNotification> = component.ngrxOnRunEffects(mockResolvedEffects$);
+			expect(component.actions$.pipe).toHaveBeenCalledTimes(1);
+
+			resolvedEffectsObservable.subscribe(mockObserver);
+
+			(<Spy>mockObserver.next).calls.reset();
+			// stop the effects
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effect.spec.ts]: This comment is not valid anymore since we saw that we could not make it work with a "stopping" action so you can remove it

File: packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effect.spec.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+			actions$.next(new StarkInitializeSession(<any>{})); // stopping action
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effect.spec.ts]: - Can you remove the comment `// stopping action`? because it is no longer valid
- Can you replace this line by these so that it is clear that it doesn't matter which action is passed the Effects will be stopped anyways?:

```
actions$.next("dummy acton");
actions$.next("another dummy acton");
```

File: packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effect.spec.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+			mockResolvedEffectsSubject.next("this should never be emitted");
+
+			expect(mockObserver.next).not.toHaveBeenCalled();
+			expect(mockObserver.error).not.toHaveBeenCalled();
+			expect(mockObserver.complete).not.toHaveBeenCalled();
+		});
+
+		it("Should use the effects", () => {
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effect.spec.ts]: - Can you add more info in the description... "should run the effects when the option xxxx is false"?
- Can you add another `it` for the case when the option is `undefined`?

File: packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effect.spec.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+			const mockResolvedEffectsSubject: Subject<any> = new Subject<any>();
+			const mockResolvedEffects$: Observable<any> = mockResolvedEffectsSubject.asObservable();
+
+			const resolvedEffectsObservable: Observable<EffectNotification> = component.ngrxOnRunEffects(mockResolvedEffects$);
+			expect(component.actions$.pipe).toHaveBeenCalledTimes(1);
+
+			resolvedEffectsObservable.subscribe(mockObserver);
+
+			// start the effects
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effect.spec.ts]: This comment is not valid anymore since we saw that we could not make it work with a "stopping" action so you can remove it

File: packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effects.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		@Inject(STARK_SESSION_SERVICE) public sessionService: StarkSessionService,
+		@Optional()
+		@Inject(STARK_SESSION_UI_CONFIG)
+		public starkSessionUiConfig: StarkSessionUiConfig,
+		@Inject(MatDialog) public dialogService: MatDialog
+	) {}
+
+	/**
+	 * This method is used to display a warning session timeout dialog.
+	 * When the dialog is closed, if the "keep-logged" string is received as a result, it means that the user sould keep logged
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/effects/session-timeout-warning.effects.ts]: Typo: "sould" -> "should"

File: packages/stark-ui/src/modules/session-ui/entities/stark-session-ui-config.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+ * The InjectionToken version of the config name
+ */
+export const STARK_SESSION_UI_CONFIG: InjectionToken<StarkSessionUiConfig> = new InjectionToken<StarkSessionUiConfig>(
+	"StarkSessionUiConfig"
+);
+
+/**
+ * Definition of the configuration object for the Stark Session service
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/entities/stark-session-ui-config.ts]: "Stark Session UI module"

File: packages/stark-ui/src/modules/session-ui/entities/stark-session-ui-config.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+ */
+export class StarkSessionUiConfig {
+	/**
+	 * Router state for the Login page where the user can choose a profile and use it to impersonate himself as someone else.
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/entities/stark-session-ui-config.ts]: ??

File: packages/stark-ui/src/modules/session-ui/entities/stark-session-ui-config.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+	 * This state is only used in DEVELOPMENT.
+	 */
+	public enableTimeoutWarningDialogEffects?: boolean;
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/entities/stark-session-ui-config.ts]: As we discussed, can you rename this option to "timeoutWarningDialogDisabled" so we keep consistency with the way the options in `stark-app-config` are named?

File: packages/stark-ui/src/modules/session-ui/session-ui.module.ts [hunk 10]
```diff
@@ -40,31 +53,25 @@ export class StarkSessionUiModule {
...
+		parentModule: StarkSessionUiModule
 	) {
 		if (parentModule) {
 			throw new Error("StarkSessionUiModule is already loaded. Import it in the AppModule only");
 		}
-
-		const english: StarkLocale = { languageCode: "en", translations: translationsEn };
-		const french: StarkLocale = { languageCode: "fr", translations: translationsFr };
-		const dutch: StarkLocale = { languageCode: "nl", translations: translationsNl };
-		mergeUiTranslations(translateService, english, french, dutch);
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/session-ui.module.ts]: I think this is why the translations are not working anymore in the Session UI module :p    Maybe a rebase problem?

File: showcase/src/app/app.module.ts [hunk 11]
```diff
@@ -235,7 +235,9 @@ export const metaReducers: MetaReducer<State>[] = ENV !== "production" ? [logger
 			position: "top right",
 			actionClasses: []
 		}),
-		StarkSessionUiModule.forRoot(),
+		StarkSessionUiModule.forRoot({
+			enableTimeoutWarningDialogEffects: true
```
Comment (christophercr) [inline showcase/src/app/app.module.ts]: You can revert this change when you rename this option to "timeoutWarningDialogDisabled" as I mentioned in my comments above ;)

File: packages/stark-ui/src/modules/session-ui/session-ui.module.ts [hunk 12]
```diff
@@ -16,22 +15,41 @@ import { translationsFr } from "./assets/translations/fr";
...
+		StarkSessionTimeoutWarningDialogComponent
 	],
-	exports: [StarkLoginPageComponent, StarkPreloadingPageComponent, StarkSessionExpiredPageComponent, StarkSessionLogoutPageComponent],
 	imports: [
 		CommonModule,
 		UIRouterModule.forChild({
 			states: SESSION_UI_STATES
 		}),
+		StarkSessionModule,
+		StarkLoggingModule,
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/session-ui.module.ts]: Please remove these two modules otherwise we are creating a hard link between the different modules of Stark.

Our goal is to have soft links, meaning that we just require the services (which the developer can provide by his own) but not the Stark modules themselves.

File: packages/stark-ui/src/modules/session-ui/entities/stark-session-ui-config.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+export const STARK_SESSION_UI_CONFIG: InjectionToken<StarkSessionUiConfig> = new InjectionToken<StarkSessionUiConfig>(
+	"StarkSessionUiConfig"
+);
+
+/**
+ * Definition of the configuration object for the Stark Session UI module
+ */
+export class StarkSessionUiConfig {
+	/**
+	 * If the warning dialog effect is disabled or not
```
Comment (christophercr) [inline packages/stark-ui/src/modules/session-ui/entities/stark-session-ui-config.ts]: Let's expand this description so that it is clear we use an effect for this. What about this?

"Whether the warning dialog should be displayed or not. If true, then the Ngrx Effects in charge of displaying the dialog won't be executed."

File: docs/TIMEOUT_WARNING.md [hunk 14]
```diff
@@ -0,0 +1,40 @@
+# Session Timeout Warning configuration
+
+Stark provides a nice feature when it comes to session expiration handling: in case the user's session is about to end due
+to inactivity (the user is idle for some time) and you want to warn him about this, a dialog will be displayed.
+
+This is what the `SessionTimeoutWarningDialogComponent` does and also asks the user if its session should be kept alive.
```
Comment (christophercr) [inline docs/TIMEOUT_WARNING.md]: Let's add this in the same line to explain how we do this ;)

"This warning dialog is displayed by an NGRX Effect that triggers when a `StarkSessionActionTypes.SESSION_TIMEOUT_COUNTDOWN_START` action is dispatched (by the Session service when the user becomes idle)."

File: packages/stark-ui/src/modules/session-ui/entities/stark-session-ui-config.ts (hunk 13 above)
Comment (Mallikki) [reply packages/stark-ui/src/modules/session-ui/entities/stark-session-ui-config.ts]: much better

File: packages/stark-ui/src/modules/session-ui/assets/translations/fr.ts [hunk 15]
```diff
@@ -22,6 +22,13 @@ export const translationsFr: object = {
 		SESSION_LOGOUT: {
 			TITLE: "Déconnecté",
 			LOGIN: "Connexion"
+		},
+		SESSION_TIMEOUT: {
+			SECONDS: " secondes.",
+			STAY_CONNECTED: "Rester connecté",
+			TITLE: "Avertissement d'expiration de session",
+			WILL_EXPIRE_IN: "Votre session va expirer dans ",
+			WISH_STAY_CONNECTED: "Voulez-vous rester connecté ?"
```
[the remark this answers is not in the document]
Comment (SuperITMan) [reply packages/stark-ui/src/modules/session-ui/assets/translations/fr.ts]: In French, we have whitespace 😊

[the remark this answers is not in the document]
Comment (christophercr) [reply packages/stark-ui/src/modules/session-ui/assets/translations/fr.ts]: WTF!! 😮

[the remark this answers is not in the document]
Comment (Mallikki) [reply packages/stark-ui/src/modules/session-ui/assets/translations/fr.ts]: In fact, if I remember well my typing lessons (yep, I've had that, boring as hell), the convention is not to put a white space before, to treat it like a regular point. Word, OpenOffice, etc, always try to put one before, but it is not the official convention ^^ So I do think Christopher is right, I'll remove it, my bad :(

[the remark this answers is not in the document]
Comment (christophercr) [reply packages/stark-ui/src/modules/session-ui/assets/translations/fr.ts]: Ok, I have no idea about this so you can do what think it is best 😉
